[ONNX] Cover 'undiscoverable' ops 'torch.ops.aten'

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99667
* __->__ #99682

Due to https://github.com/pytorch/pytorch/issues/99681, ops supported by torchlib in `function_dispatcher` may not get
registered. This PR works around it by doing reverse look up.
Also fixes aten signature for `log_sigmoid`, which appears to be an outlier with unique naming style.

BowenBao committed:
[ONNX] Cover 'undiscoverable' ops 'torch.ops.aten'

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -159,8 +160,19 @@ def _create_op_overload_to_exporter_key_table() -> (
+    # Some ops in `torch.ops.aten` are not discoverable through `dir(torch.ops.aten)`,
+    # but retrievable via explicit lookup.
+    # https://github.com/pytorch/pytorch/issues/99681
+    # This is a workaround to make sure we register ONNX symbolic functions for these.
```
Does this way help us find aten_sym_size?

BowenBao replied:
Great question, it does :)

BowenBao replied:
```
>>> import torch
>>> "sym_size" in torch.ops.aten
False
>>> torch.ops.aten.sym_size
<OpOverloadPacket(op='aten.sym_size')>
```

justinchuby replied:
Should we move sym_size to torchlib

justinchuby replied:
I am happy to include as many as we see

BowenBao replied:
Curious if it has opinfo. I wonder if it means higher maintenance cost than rest of 'normal' ops. I don't have strong opinion either way.

titaiwangms replied:
What does it mean "moving to torchlib"? Isn't it in onnx-script core.py?

BowenBao replied:
thanks @titaiwangms, case closed!

titaiwangms replied:
😨

justinchuby replied:
Ahh sorry I mistook it for getitem

BowenBao committed:
Update on "[ONNX] Cover 'undiscoverable' ops 'torch.ops.aten'"


Due to https://github.com/pytorch/pytorch/issues/99681, ops supported by torchlib in `function_dispatcher` may not get
registered. This PR works around it by doing reverse look up.
Also fixes aten signature for `log_sigmoid`, which appears to be an outlier with unique naming style.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -159,8 +160,19 @@ def _create_op_overload_to_exporter_key_table() -> (
+    # https://github.com/pytorch/pytorch/issues/99681
+    # This is a workaround to make sure we register ONNX symbolic functions for these.
+    onnx_supported_aten_lookup_table = [
+        k.split("::")[1] for k in _ATENLIB_FUNCTIONS.keys() if k.startswith("aten::")
```
prims too?

justinchuby replied:
nvm

justinchuby commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -159,8 +160,19 @@ def _create_op_overload_to_exporter_key_table() -> (
+        if op_namespace is torch.ops.aten:
+            attr_names += onnx_supported_aten_lookup_table
+        for attr_name in attr_names:
             op_overload_packet = getattr(op_namespace, attr_name)
```
Should we handle when attr_name is not in op_namespace?

BowenBao replied:
No, it will crash. Can be an assertion but I was being lazy since we have hand written mapping table and this is caught at loading for now.

BowenBao replied:
At this point I won't update in this PR, unless a merge conflict happens with #99284 :)

justinchuby replied:
Cool. I will wait for this to land first. Although there could be an issue when torch decides to remove ops

BowenBao replied:
Oh interesting, are you thinking about relaxing the constraint? My thought was to reveal this as soon as possible as CI breaks. Assertion just makes it explicit.

justinchuby replied:
I am undecided. Letting it break makes sense too considering we want to know asap when something like this happens

BowenBao commented:
@pytorchbot merge
